src/sentry/models/releasefile.py

```diff
@@ -118,3 +122,67 @@ def clear_old_entries(self):
 
 
 ReleaseFile.cache = ReleaseFileCache()
+
+
+class ReleaseArchive:
+    """ Read-only view of uploaded ZIP-archive of release files """
+
+    def __init__(self, fileobj: IO):
+        self._fileobj = fileobj
+        self._zip_file = zipfile.ZipFile(self._fileobj)
+        self.manifest = self._read_manifest()
+        files = self.manifest.get("files", {})
+
+        self._entries_by_url = {entry["url"]: (path, entry) for path, entry in files.items()}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc, value, tb):
+        self._zip_file.close()
+        self._fileobj.close()
+
+    def read(self, filename: str) -> bytes:
+        return self._zip_file.read(filename)
+
+    def _read_manifest(self) -> dict:
+        manifest_bytes = self.read("manifest.json")
+        return json.loads(manifest_bytes.decode("utf-8"))
+
+    def get_file_by_url(self, url: str) -> Tuple[IO, dict]:
+        """Return file-like object and headers.
+
+        The caller is responsible for closing the returned stream.
+
+        May raise ``KeyError``
+        """
+        filename, entry = self._entries_by_url[url]
+        return self._zip_file.open(filename), entry.get("headers", {})
+
+    def extract(self) -> TemporaryDirectory:
+        """Extract contents to a temporary directory.
+
+        The caller is responsible for cleanup of the temporary files.
+        """
+        temp_dir = TemporaryDirectory()
+        safe_extract_zip(self._fileobj, temp_dir.name, strip_toplevel=False)
+
+        return temp_dir
+
+
+def merge_release_archives(archive1: ReleaseArchive, archive2: ReleaseArchive, target: IO):
+    """Fields in archive2 take precedence over fields in archive1. """
+    merged_manifest = dict(archive1.manifest, **archive2.manifest)
+    files1 = archive1.manifest.get("files", {})
+    files2 = archive2.manifest.get("files", {})
+
+    merged_manifest["files"] = dict(files1, **files2)
+
+    with zipfile.ZipFile(target, mode="w", compression=zipfile.ZIP_DEFLATED) as zip_file:
+
+        for filename in files2.keys():
+            zip_file.writestr(filename, archive2.read(filename))
+        for filename in files1.keys() - files2.keys():
+            zip_file.writestr(filename, archive1.read(filename))
```

Comment:
A few points and questions on that function:
1. What does `writestr` do when the file already exists? Upon quick search, I was not able to find an answer to this in `zipfile` docs. 
2. Could we open the _larger_ file to add missing files rather than recreating a ZIP from scratch?
3. Let's double-check that we're using the same compression level as in sentry-cli.

Reply:
1. I remember now, calling `writestr` twice creates another entry with the same name. Also, there's no way to delete an entry from the `ZipFile`, so
2. recreating from scratch seemed the only way forward.
3. They both seem to use zlib compression (although symbolic also has a `bzip2` feature?). Unfortunately, we cannot set the `compression_level` parameter in Python 3.6.